hierarchicalforecast/core.py

```diff
@@ -97,99 +103,142 @@ def __init__(self,
         self.insample = any([method.insample for method in reconcilers])
     
     def _prepare_fit(self,
-                     Y_hat_df: pd.DataFrame,
-                     S_df: pd.DataFrame,
-                     Y_df: Optional[pd.DataFrame],
+                     Y_hat_df: FrameT,
+                     S_df: FrameT,
+                     Y_df: Optional[FrameT],
                      tags: Dict[str, np.ndarray],
                      level: Optional[List[int]] = None,
                      intervals_method: str = 'normality',
-                     sort_df: bool = True):
+                     sort_df: bool = True,
+                     id_col: str = "unique_id",
+                     time_col: str = "ds", 
+                     target_col: str = "y",                      
+                     ):
         """
         Performs preliminary wrangling and protections
         """
+        Y_hat_df_nw = nw.from_native(Y_hat_df)
+        S_df_nw = nw.from_native(S_df)
+        Y_df_nw = None
+        self.native_namespace = nw.get_native_namespace(Y_hat_df_nw)
+
         #-------------------------------- Match Y_hat/Y/S index order --------------------------------#
-        if sort_df:
-            Y_hat_df = Y_hat_df.reset_index()
-            Y_hat_df.unique_id = Y_hat_df.unique_id.astype('category')
-            Y_hat_df.unique_id = Y_hat_df.unique_id.cat.set_categories(S_df.index)
-            Y_hat_df = Y_hat_df.sort_values(by=['unique_id', 'ds'])
-            Y_hat_df = Y_hat_df.set_index('unique_id')
-
-            if Y_df is not None:
-                Y_df = Y_df.reset_index()
-                Y_df.unique_id = Y_df.unique_id.astype('category')
-                Y_df.unique_id = Y_df.unique_id.cat.set_categories(S_df.index)
-                Y_df = Y_df.sort_values(by=['unique_id', 'ds'])
-                Y_df = Y_df.set_index('unique_id')
-
-            S_df.index = pd.CategoricalIndex(S_df.index, categories=S_df.index)
+        # TODO: This is now a bit slow as we always sort.
+        S_df_nw = S_df_nw.with_columns(nw.from_dict({f"{id_col}_id": np.arange(len(S_df_nw))}, 
+                                                native_namespace=self.native_namespace)[f"{id_col}_id"])
+        Y_hat_df_nw = Y_hat_df_nw.join(S_df_nw[[id_col, f"{id_col}_id"]], on=id_col, how='left')
+        Y_hat_df_nw = Y_hat_df_nw.sort(by=[f"{id_col}_id", time_col])
+        Y_hat_df_nw = Y_hat_df_nw.drop(f"{id_col}_id")
+
+        if Y_df is not None:
+            Y_df_nw = nw.from_native(Y_df)
```

Comment:
It's clearer to define this here
```suggestion
        if Y_df is None:
            Y_df_nw = None
        else:
            Y_df_nw = nw.from_native(Y_df)
```